runtime/v1/shim/client/client.go

```diff
@@ -324,17 +324,18 @@ func (c *Client) signalShim(ctx context.Context, sig syscall.Signal) error {
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
-	case <-c.waitForExit(pid):
+	case <-c.waitForExit(pid, sig):
 		return nil
 	}
 }
 
-func (c *Client) waitForExit(pid int) <-chan struct{} {
+func (c *Client) waitForExit(pid int, sig syscall.Signal) <-chan struct{} {
 	c.exitOnce.Do(func() {
 		for {
-			// use kill(pid, 0) here because the shim could have been reparented
+			// use kill(pid, sig) here because the shim could have been reparented
 			// and we are no longer able to waitpid(pid, ...) on the shim
-			if err := unix.Kill(pid, 0); err == unix.ESRCH {
+			// we use sig again instead of 0 here because we have seen a single kill not work sometimes
```

Comment:
If `kill` doesn't work at this moment, could the caller retry kill shim again? It is reasonable to do retry if failed at first time. 

For the mentioned case, if `kill -9` fails at the first time, can you kill it by manually? And could it be possible reused pid here (same pid but it is for different container)?

Reply:
could you mind to check `wchan` result from `ps axo pid,cmd,wchan` if the shim is still hanging there?

Reply:
@fuweid yea, we have seen retrying the kill succeed when the first kill fails. 

The manual kill succeeds.

```
31745 containerd-shim -namespace  futex_wait_queue_me
```
is the output for one such shim.

Reply:
Thanks for the information. And the shim is the same shim? I mean the pid can be reused.

Reply:
yes i confirmed the shim is the same based on the container id it was associated with.